src/d.sea-of-nodes.md

```diff
@@ -0,0 +1,327 @@
+          id: { type: 'Identifier', name: 'i' },
+          init: { type: 'Literal', value: 0, raw: '0' } } ],
+     kind: 'var' },
+
+  //
+  // `i < arr.length`
+  //
+  test:
+   { type: 'BinaryExpression',
+     operator: '<',
+     left: { type: 'Identifier', name: 'i' },
+     right:
+      { type: 'MemberExpression',
+        computed: false,
+        object: { type: 'Identifier', name: 'arr' },
+        property: { type: 'Identifier', name: 'length' } } },
+
+  //
+  // `i++`
+  //
+  update:
+   { type: 'UpdateExpression',
+     operator: '++',
+     argument: { type: 'Identifier', name: 'i' },
+     prefix: false },
+
+  //
+  // `arr[i] += 1;`
+  //
+  body:
+   { type: 'ExpressionStatement',
+     expression:
+      { type: 'AssignmentExpression',
+        operator: '+=',
+        left:
+         { type: 'MemberExpression',
+           computed: true,
+           object: { type: 'Identifier', name: 'arr' },
+           property: { type: 'Identifier', name: 'i' } },
+        right: { type: 'Literal', value: 1, raw: '1' } } } }
+```
+
+This JSON could also be visualized:
+![AST][7]
+
+This is a tree, so it is very natural to traverse it from the top to the bottom,
+generating the machine code as we visit the AST nodes. The problem with this
+approach is that the information about variables is very sparse, and is spread
+through the different tree nodes.
+
+To safely move the length lookup out of the loop we need to know that the array
+length does not change between the loop's iterations. Humans can do it easily
+just by looking at the source code, but the compiler needs to do lots of work to
+figure this out of the AST.
+
+As many other compiler problems, this one is commonly solved by introducing one
+more intermediate representation layer. What's needed to optimize this
+particular case is a thing called data-flow graph. Instead of talking about
+syntax-entities (like `for loop`s, `expressions`, ...), we should talk about the
+data itself (read variables values), and how it changes through the program.
+
+## Data-flow Graph
+
+In our particular example, the data we are interested in is the value of
+variable `arr`. We want to be able to easily observe all uses of it to verify
+that there are no out-of-bands accesses and other length-changing uses of that
+array.
+
+This is accomplished by introducing def-use (definition and uses) relationship
+between the different data values, and it means that the value is declared once
+(_node_), and is used somewhere to create new values (_edge_). Obviously,
+connecting different values together will form a **data-flow graph** like this:
+
+![Data-flow Graph][8]
+
+Note the red `array` box in this vast graph. The solid arrows going out of it
+represent uses of this value. By iterating them compiler can tell that `array`
+value is used at:
+
+- `loadArrayLength`
+- `checkRange`
+- `load`
+- `store`
+
+It could also follow back the dashed lines (these lines are very important, and
```

Comment:
```diff
- It could also follow back the dashed lines (these lines are very important, and we will discuss them a bit later) going to `checkRange` to verify that both `load` and `store` are "guarded" by it, and won't cause array length to change. Compiler concludes that it is safe to move `loadArrayLength` out of the loop.
+ The compiler could also follow the dashed lines (these lines are very important, and we will discuss them a bit later), back to `checkRange` to verify that both `load` and `store` are "guarded", i.e. the array's uses won't cause its length to change. Thus the compiler will conclude that it is safe to move `loadArrayLength` out of the loop.
```